src/notification.c

```diff
@@ -240,6 +240,136 @@ check_value_type (const char *key,
   return FALSE;
 }
 
+static void
+markup_parser_text (GMarkupParseContext  *context,
+                    const gchar          *text,
+                    gsize                 text_len,
+                    gpointer              user_data,
+                    GError              **error)
+{
+  GString *composed = user_data;
+  gchar *str = (gchar *) text;
+
+  while (*str)
+    {
+      gunichar c = g_utf8_get_char (str);
+      gboolean needs_space = FALSE;
+
+      while (g_unichar_isspace (c) && *str)
+        {
+          needs_space = composed->len > 0 && composed->str[composed->len - 1] != ' ';
+
+          str = g_utf8_next_char (str);
+
+          if (*str)
+            c = g_utf8_get_char (str);
+        }
+
+      if (needs_space)
+        g_string_append_c (composed, ' ');
+
+      if (*str)
+        {
+          g_string_append_unichar (composed, c);
+          str = g_utf8_next_char (str);
+        }
+    }
+}
+
+static void
+markup_parser_start_element (GMarkupParseContext *context,
+                             const gchar         *element_name,
+                             const gchar         **attribute_names,
+                             const gchar         **attribute_values,
+                             gpointer             user_data,
+                             GError              **error)
+{
+  GString *composed = user_data;
+
+  if (strcmp (element_name, "b") == 0)
+    {
+      g_string_append_len (composed, "<b>", -1);
+    }
+  else if (strcmp (element_name, "i") == 0)
+    {
+      g_string_append_len (composed, "<i>", -1);
+    }
+  else if (strcmp (element_name, "a") == 0)
+    {
+      gsize i = 0;
+
+      while (attribute_names[i])
+        {
+          if (strcmp (attribute_names[i], "href") == 0)
+            {
+              g_string_append_printf (composed, "<a href='%s'>", attribute_values[i]);
+              break;
+            }
+
+          i++;
+        }
+    }
+}
+
+static void
+markup_parser_end_element (GMarkupParseContext *context,
+                           const gchar         *element_name,
+                           gpointer             user_data,
+                           GError              **error)
+{
+  GString *composed = user_data;
+
+  if (strcmp (element_name, "b") == 0)
+    g_string_append_len (composed, "</b>", -1);
+  else if (strcmp (element_name, "i") == 0)
+    g_string_append_len (composed, "</i>", -1);
+  else if (strcmp (element_name, "a") == 0)
+    g_string_append_len (composed, "</a>", -1);
+}
+
+static const GMarkupParser markup_parser = {
```

Comment:
I think you know, but this parser doesn't make sure elements are balanced, which it probably should...

Reply:
That's the reason why I added the `<markup></markup>` wrapper so if the elements aren't closed correctly this will fail. I got the idea from pango.

Reply:
Ah, neat trick but should be in a comment then.

Reply:
Actually it does make sure elements are balanced. Checking the docs again https://docs.gtk.org/glib/method.MarkupParseContext.end_parse.html it says that. The `<markup></markup>` is needed because the MarkupParse expects the markup to start with an element.

Reply:
I added a comment for the reason why `<markup>` is added.